mne/io/edf/edf.py

```diff
@@ -188,100 +159,179 @@ def _read_segment(self, start=0, stop=None, sel=None, verbose=None,
         picks = [stim_channel, tal_channel]
         offsets[picks] = 0
 
-        with open(self.info['filename'], 'rb') as fid:
+        # set up output array
+        data_shape = (len(sel), stop - start)
+        if isinstance(data_buffer, np.ndarray):
+            if data_buffer.shape != data_shape:
+                raise ValueError('data_buffer has incorrect shape')
+            data = data_buffer
+        else:
+            data = np.empty(data_shape, dtype=float)
+
+        read_size = blockstop - blockstart
+        this_data = np.empty((len(sel), buf_len))
+        """
+        Consider this example:
+
+        tmin, tmax = (3, 27)
+        read_size = 30
+        buf_len = 10
+        sfreq = 1.
+
+                        +---------+---------+---------+
+        File structure: |  buf0   |   buf1  |   buf2  |
+                        +---------+---------+---------+
+        File time:      0        10        20        30
+                        +---------+---------+---------+
+        Requsted time:    2                       27
+
+                        |                             |
+                    blockstart                    blockstop
+                          |                        |
+                        start                    stop
+
+        We need 27 - 2 = 25 samples (per channel) to store our data, and
+        we need to read from 3 buffers (30 samples) to get all of our data.
+
+        On all reads but the first, the data we read starts at
+        the first sample of the buffer. On all reads but the last,
+        the data we read ends on the last sample of the buffer.
+
+        We call this_data the variable that stores the current buffer's data,
+        and data the variable that stores the total output.
+
+        On the first read, we need to do this::
+
+            >>> data[0:buf_len-2] = this_data[2:buf_len]
+
+        On the second read, we need to do::
+
+            >>> data[1*buf_len-2:2*buf_len-2] = this_data[0:buf_len]
+
+        On the final read, we need to do::
+
+            >>> data[2*buf_len-2:3*buf_len-2] = this_data[0:buf_len-3]
```

Comment:
this example is great, really clarifies the reading, but should it be:

`data[2*buf_len-2:3*buf_len-2-3]`?